patterns/strategy.md

```diff
@@ -0,0 +1,171 @@
+# Strategy (aka Policy)
+
+## Description
+
+The [Strategy design pattern](https://en.wikipedia.org/wiki/Strategy_pattern)
+is a technique that enables separation of concerns in object-oriented software design.
+It also allows to decouple software modules through [Dependency Inversion](https://en.wikipedia.org/wiki/Dependency_inversion_principle).
+The basic idea behind the Strategy pattern is that, given an algorithm solving a particular problem,
+we define only the skeleton of the algorithm at an abstract level and
+we delegate the specific algorithm’s implementation (or some of its parts) to another class or method.
+
+In this way, a client using the algorithm may choose a specific implementation, while the general algorithm workflow remains the same.
+In other words, the abstract specification of the class does not depend on the specific implementation of the derived class,
+but specific implementation must adhere to the abstract specification.
+This is why we call it "Dependency Inversion".
+
+## Motivation
+
+Imagine we are working on a project that generates reports every month.
+We need the reports to be generated in different formats (strategies), e.g.,
+in `JSON` or `Plain Text` formats.
+But things vary over time and we don't know what kind of requirement we may get in the future.
+For example, we may need to generate our report in a completly new format,
+or just modify one of the existing formats.
+
+## Example
+
+In this example our invariants (or abstractions) are `Context`, `Formatter`, and `Report`,
+while `Text` and `Json` are our strategy structs.
+These strategies have to implement the `Formatter` trait.
+
+```rust
+use std::collections::HashMap;
+type Data = HashMap<String, u32>;
+
+trait Formatter {
+    fn run(&self, data: &Data) -> String;
+}
+
+struct Report;
+
+impl Report {
+    fn generate<T: Formatter>(g: T) -> String {
```

Comment:
Ah, like I mentioned before. It's just an example. Since we are in rust patterns, I think we should show a good pattern since others will be learning from here. I still think taking in `&mut String` should be used here, at least others not knowing this will be more aware about it, it won't make it much complicated right?

Reply:
@fade2black Seems like this is the last review comment that we could address before merging, any thoughts on it?

Reply:
@simonsan  I am on it.

Reply:
@simonsan @pickfire Fixed.

Reply:
I think that taking the string as mut is not always the best approach.
The rust library is full of functions that returns strings as output.
Taking it as mut gives you flexibility, but it is more error prone and less readable. It depends on your priorities.
Anyway, I don't think it matters that much, so it's fine like this :)

Reply:
Whether it is `&mut String` or `String` as return, I don't care very much in this example. This is just an example.
Depending on use cases both `&mut String` and `String` as return would work fine. No need to waste time on [trivial things](https://en.wikipedia.org/wiki/Law_of_triviality).